crates/move-stackless-bytecode/src/quantifier_iterator_analysis.rs

```diff
@@ -0,0 +1,226 @@
+            _ => false,
+        }
+    }
+
+    fn is_trace_local(&self, bc: &Bytecode) -> bool {
+        match bc {
+            Bytecode::Call(_, _, op, _, _) => matches!(op, Operation::TraceLocal(_)),
+            _ => false,
+        }
+    }
+
+    fn is_searched_fn(&self, bc: &Bytecode, qid: QualifiedId<FunId>) -> bool {
+        match bc {
+            Bytecode::Call(_, _, operation, _, _) => {
+                match operation {
+                    Operation::Function(mod_id,fun_id, _) => {
+                        qid == mod_id.qualified(*fun_id)
+                    },
+                    _ => false
+                }
+            },
+            _ => false,
+        }
+    }
+
+    fn validate_function_pattern_requirements(&self, env: &GlobalEnv, targets: &FunctionTargetsHolder, qid: QualifiedId<FunId>) -> bool {
+        let func_env = env.get_function(qid);
+        let data = targets.get_data(&qid, &FunctionVariant::Baseline).unwrap();
+
+        if !no_abort_analysis::get_info(data).does_not_abort && !targets.is_abort_check_fun(&qid) {
+            env.diag(
+                Severity::Error,
+                &func_env.get_loc(),
+                "Quantifier function should not abort",
+            );
+
+            return true;
+        }
+
+        if !deterministic_analysis::get_info(data).is_deterministic {
+            env.diag(
+                Severity::Error,
+                &func_env.get_loc(),
+                "Quantifier function should be deterministic",
+            );
+
+            return true;
+        }
+
+        return false;
+    }
+
+    pub fn find_macro_patterns(&self, env: &GlobalEnv, targets: &FunctionTargetsHolder, pattern: &QuantifierPattern, bc: &Vec<Bytecode>) -> Vec<Bytecode> {
+        let chain_len = 5;
+        if bc.len() < chain_len {
+            return bc.to_vec();
+        }
+
+        for i in 0..bc.len() - (chain_len - 1) {
+            if 
+                self.is_searched_fn(&bc[i], pattern.start_qid) &&
+                self.is_trace_local(&bc[i + 1]) &&
+                self.is_fn_call(&bc[i + 2]) && 
+                self.is_destroy(&bc[i + 3]) &&
+                self.is_searched_fn(&bc[i + 4], pattern.end_qid) 
+            {
+                let (attr_id, _, srcs_base, callee_id, type_params) = self.extract_fn_call_data(&bc[i + 2]);
+                let (_, _, srcs_vec, _, _) = self.extract_fn_call_data(&bc[i]);
+                let (_, dsts, _, _, _) = self.extract_fn_call_data(&bc[i + 4]);
+
+                if self.validate_function_pattern_requirements(env, targets, callee_id) {
+                    return bc.to_vec();
+                }
+
+                let mut new_bc = bc.clone();
+                let new_bc_el = Bytecode::Call(
+                    attr_id,
+                    dsts,
+                    Operation::Quantifier(pattern.quantifier_type, callee_id, type_params),
+                    if srcs_vec.len() > 0 { srcs_vec } else { srcs_base },
```

Comment:
`srcs_vec` is empty only for `forall`/`exists`, and in that case there aren't any sources, so it's always `srcs_vec` (nit: rename to `srcs`)